### PR TITLE
[exporterhelper] enable Persistent Queue feature by default

### PR DESCRIPTION
**Pull request by swiatekm-sumo**
**Description:**
Persistent Queue is a feature that currently requires "enable_unstable" tag provided to the the build to enable it. We have been using it in [our distro](https://github.com/SumoLogic/sumologic-otel-collector/) for several months already and it looks stable. It seems that others are using this capability as well (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9327#issuecomment-1135494065)

This change enables it by default, which also simplifies the build process a little bit.

**Testing:**
Added some tests to cover the queue creation logic and the requeue feature. They're a bit awkward, and should probably be refactored once the requeue feature is exposed in config.

Originally posted by @pmm-sumo [here](https://github.com/open-telemetry/opentelemetry-collector/pull/5457), see the review discussion therein for justification for some of the choices made.

### CODE REVIEW

**Comment by swiatekm-sumo** [comment]
I've rebased this post 0.56.0 and fixed the remaining minor issues the CI got stuck on.

File: exporter/exporterhelper/queued_retry_inmemory.go
Comment (bogdandrutu) [inline exporter/exporterhelper/queued_retry_inmemory.go:48]: I don't like this config, is too complex for no good reason. I see no point in supporting `persistent_storage_enabled` instead of simply asking for the exact name.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Would like to agree on the config first, since that is the most important part.

File: exporter/exporterhelper/queued_retry_inmemory.go
Comment (swiatekm-sumo) [reply exporter/exporterhelper/queued_retry_inmemory.go:48]: The current state was an attempt to keep the config backwards-compatible, as this whole change was originally just about enabling the queue in the build by default. If we're going to break it, then I agree with @bogdandrutu - just make the user specify the storage explicitly and be done with it, looks like the best option in the long term.

If we do that though, I think the changelog should mention the change explicitly.

Comment (bogdandrutu) [reply exporter/exporterhelper/queued_retry_inmemory.go:48]: I would not call this a breaking change since this was an experimental feature.

Comment (swiatekm-sumo) [reply exporter/exporterhelper/queued_retry_inmemory.go:48]: It's not technically a breaking change, but it's nonetheless a feature being used in the real world, and I think it'd be polite to call it out in the changelog to make it easier for users to figure out what broke their config after an upgrade.

Either way, I'll make the change in this PR.

Comment (bogdandrutu) [reply exporter/exporterhelper/queued_retry_inmemory.go:48]: FYI: Waiting for the change.

Comment (swiatekm-sumo) [reply exporter/exporterhelper/queued_retry_inmemory.go]: Done, sorry for the delay. I've added a changelog entry to breaking changes, even though this technically isn't one.

File: CHANGELOG.md
Comment (tigrannajaryan) [inline CHANGELOG.md:16]: This is a bit misleading. We don't enable persistent queue by default. We just make the feature available in the default build.

File: Makefile
Comment (tigrannajaryan) [inline Makefile:61]: Why is this removed? What if we have other unstable features to test?

File: exporter/exporterhelper/queued_retry_inmemory.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/queued_retry_inmemory.go:121]: Ideally this PR should do one thing: enable the feature. Instead I see modifications of the logic. Can we split the PR into 2 parts:
1. Change whatever logic needs to change and whatever refactoring is necessary. Keep the persistent queue feature experimental.
2. Make persistent queue feature part of stable build.
Otherwise I have a hard time understanding what changes are needed for enabling the feature, which changes are refactorings which don't change the behavior and what changes change the behavior.

File: CHANGELOG.md
Comment (bogdandrutu) [inline CHANGELOG.md:7]: Please comment, this happens only if you enable the experimental feature xyz...

File: Makefile
Comment (bogdandrutu) [reply Makefile:61]: I think we should go with feature gates for unstable features from now on, this is an old unstable feature before the gates.

File: exporter/exporterhelper/README.md
Comment (bogdandrutu) [inline exporter/exporterhelper/README.md:42]: wrong statement, they are buffered in the storage :) the storage can be s3 or nvram (whatever the extension implements).

File: exporter/exporterhelper/internal/persistent_queue.go
Comment (bogdandrutu) [inline exporter/exporterhelper/internal/persistent_queue.go:39]: why is this change necessary now and not before? if this can be "refactored" as a separate PR, please do so.

**Review by bogdandrutu — COMMENTED** [review]
Please make sure we are not doing too many things in one PR, and keep PRs as small and focus as possible.

File: exporter/exporterhelper/queued_retry_inmemory.go
Comment (swiatekm-sumo) [reply exporter/exporterhelper/queued_retry_inmemory.go:121]: Yeah, this became too complicated along the way. I'm going to split it into three PRs:
1. Make the queue logic consistent between the in-memory queue and the persistent one. This way enabling the persistent queue is just a change to the build.
2. Enable persistent queue in the build by default.
3. Configuration changes for setting the storage for the persistent queue.

Comment (swiatekm-sumo) [reply exporter/exporterhelper/queued_retry_inmemory.go:121]: The first PR: https://github.com/open-telemetry/opentelemetry-collector/pull/5764. Would you prefer to close this one right now or wait until all the changes are merged individually? @tigrannajaryan @bogdandrutu ?

**Comment by bogdandrutu** [comment]
This needs some rework.

**Comment by swiatekm-sumo** [comment]
> This needs some rework.

I'll just close it and continue submitting the individual changes as separate PRs. After #5764, I'll do the storage configuration change next.

**Pull request closed without merging**
